added support for marking on google maps

shreyas-satish opened the pull request.

shreyas-satish commented:
Sample Usage:

```python
location = MapMarkerField(__("Location"), map_id="map", longitude_id="longitude", latitude_id="latitude")
latitude = wtforms.DecimalField(__("Latitude"), places=None,
        validators=[wtforms.validators.Optional(), wtforms.validators.NumberRange(-90, 90)])
longitude = wtforms.DecimalField(__("Longitude"), places=None,
        validators=[wtforms.validators.Optional(), wtforms.validators.NumberRange(-180, 180)])

```

jace commented on baseframe/templates/baseframe.html:
```diff
@@ -109,7 +109,7 @@
     <script type="text/javascript" src="{{ ASSET_URL }}"></script>
   {%- endassets -%}
-
+  <script type="text/javascript" src='http://maps.google.com/maps/api/js?sensor=false&libraries=places'></script>
```
`&` -> `&amp;`

Also, this shouldn't be included unless there's a MapField in the form. Look for this in forms.html.

jace commented on baseframe/templates/baseframe/forms.html:
```diff
@@ -58,6 +57,8 @@
           {%- elif field.widget.input_type in ['text', 'password', 'color', 'date', 'datetime', 'datetime-local', 'email', 'month', 'number', 'search', 'tel', 'time', 'url', 'week'] or (field.type in ['SelectField', 'QuerySelectField', 'SelectMultipleField', 'QuerySelectMultipleField', 'TextAreaField', 'TextListField'] and field.widget.html_tag not in ['ul', 'ol']) %}
             {{ field | render_field_options(class="form-control field-" + field.id + " " + widget_css_class, tabindex=tabindex, autofocus=autofocus, rows=rows)}}
+          {%- elif field.type == 'MapMarkerField' %}
+            {{add_map_marker_widget(field)}}
```
`{{` and `}}` need spaces inside. Change the macro name to `map_marker_widget`.

jace commented on baseframe/templates/baseframe/forms.html:
```diff
@@ -247,4 +248,26 @@
\ No newline at end of file
+{% endmacro %}
+
+{% macro add_map_marker_widget(field) %}
```
Rename to `map_marker_widget`.

jace commented on baseframe/templates/baseframe/forms.html:
```diff
@@ -247,4 +248,26 @@
+{% endmacro %}
+
+{% macro add_map_marker_widget(field) %}
+  <style>
```
Style shouldn't be inline. `height: 20em` isn't a safe assumption. This will need styling per responsive breakpoint.

jace commented on baseframe/templates/baseframe/forms.html:
```diff
@@ -247,4 +248,26 @@
+        map_id: "{{field.map_id}}",
+        longitude_id: "{{field.longitude_id}}",
+        latitude_id: "{{field.latitude_id}}",
+        location_id: "{{field.location_id}}"
```
These three ids -- location, latitude and longitude -- should be auto-generated from the field's id. These are internal API detail anyway, the view doesn't need to know how the form works internally.

jace commented on baseframe/assets.py:
```diff
@@ -254,3 +254,8 @@
 ]}
+
+assets['locationpicker.jquery.js'][Version('0.1.12')] = ('baseframe/js/locationpicker.jquery.js')
+assets['map-marker.js'][Version(__version__)] = {'requires': [
```
Unless map-marker.js is our code, this needs upstream version number.

jace commented on baseframe/assets.py:
```diff
@@ -254,3 +254,8 @@
     'baseframe-networkbar.css==' + __version__,
 ]}
+
+assets['locationpicker.jquery.js'][Version('0.1.12')] = ('baseframe/js/locationpicker.jquery.js')
```
Depends on jQuery, so needs to be `('jquery.js', 'baseframe/js/locationpicker.jquery.js')`.

jace commented on baseframe/forms/fields.py:
```diff
@@ -503,3 +503,19 @@ def process(self, *args, **kwargs):
+    Adds a map widget that allows the user to search for a location and drag the marker on the map.
+    See the add_map_marker_widget macro in forms.html.
+    """
+    def __init__(self, label='', validators=None, attr='map', map_id='map', latitude_id='', longitude_id='', **kwargs):
```
Don't accept: `map_id`, `latitude_id`, `longitude_id`. Autogenerate all of them.

Also, you need to supply a widget that will be used if there is no JS. Create a new widget class that renders two text fields for lat and lon.

shreyas-satish commented:
> & -> \&amp;
>  {{ and }} need spaces inside. Change the macro name to map_marker_widget.
> Rename to map_marker_widget.
> Unless map-marker.js is our code, this needs upstream version number.
> Depends on jQuery, so needs to be ('jquery.js', 'baseframe/js/locationpicker.jquery.js').

Fixed

> Style shouldn't be inline. height: 20em isn't a safe assumption. This will need styling per responsive breakpoint.

Moved the map class to _layout.sass. Isn't it better to have the client specify the styling for the map widget based on the layout and baseframe just tries to provide a sane default?

> These three ids -- location, latitude and longitude -- should be auto-generated from the field's id. These are internal API detail anyway, the view doesn't need to know how the form works internally.
>  Don't accept: map_id, latitude_id, longitude_id. Autogenerate all of them.
> Also, you need to supply a widget that will be used if there is no JS. Create a new widget class that renders two text fields for lat and lon.

Pending.

> Also, this shouldn't be included unless there's a MapField in the form. Look for this in forms.html.

Using g.gmap for now.

jace commented:
> Moved the map class to _layout.sass. Isn't it better to have the client specify the styling for the map widget based on the layout and baseframe just tries to provide a sane default?

That should be in the app's CSS, not in the form.

shreyas-satish commented:
Ok, so it's no longer in the form. We'll need a default, because the map won't load otherwise.
